Docs for SCIM provisioning/deprovisioning of service accounts (#2641)
wandb/docs: platform/hosting/iam/scim.mdx

## User management

The SCIM user resource maps to W&B users and service accounts. Use these endpoints to manage users and service accounts in your organization.

> For service account concepts and UI workflows, see [Use service accounts to automate workflows](/platform/hosting/iam/service-accounts/).

Author: ```suggestion
For service account concepts and UI workflows, see [Use service accounts to automate workflows](/platform/hosting/iam/service-accounts).
```

</Tab>
</Tabs>

### Provision service account

> Creates a team-scoped or organization-scoped service account in your organization. Omit `accountType` to create a regular user instead; see [Create user](#create-user).

Author: ```suggestion
Creates a team-scoped or organization-scoped service account in your organization. Omit `accountType` to create a regular user instead. See [Create user](#create-user).
```

<Note>
Available in **Dedicated Cloud** and **Self-Managed** v0.81.0+ and in **Multi-tenant Cloud**.

- Set `userName` to the service account name. The API uses `userName` for the account's display name; `displayName` in the request body is ignored.
- `emails` are not required for service accounts.
- `modelsSeat` and `weaveRole` are not supported on create and return `400 Bad Request` if present.
- Service accounts cannot be updated with `PATCH` or `PUT`, cannot be deactivated, and cannot be assigned organization, team, or registry roles through SCIM. Create API keys in the W&B App after provisioning.
</Note>

#### Endpoint
- **URL**: `<host-url>/scim/Users`
- **Method**: POST

#### Parameters
| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|

> | `userName` | string | Yes | Unique name for the service account |

Reviewer: ```suggestion
| `userName` | string | Yes | Unique name for the service account. |
```

> | `accountType` | string | Yes | `SERVICE` for a [team-scoped service account](/platform/hosting/iam/service-accounts/#team-scoped-service-accounts), or `ORG_SERVICE` for an [organization-scoped service account](/platform/hosting/iam/service-accounts/#organization-scoped-service-accounts) |

Reviewer: ```suggestion
| `accountType` | string | Yes | `SERVICE` for a [team-scoped service account](/platform/hosting/iam/service-accounts/#team-scoped-service-accounts), or `ORG_SERVICE` for an [organization-scoped service account](/platform/hosting/iam/service-accounts/#organization-scoped-service-accounts). |
```

> | `urn:ietf:params:scim:schemas:extension:teams:2.0:User` | object | Yes | Teams extension object |

Reviewer: ```suggestion
| `urn:ietf:params:scim:schemas:extension:teams:2.0:User` | object | Yes | Teams extension object. |
```

Author: Iffy about this one but I guess at least it makes it match the others

| `defaultTeam` | string | Yes | Sub-field of the teams extension. Name of an existing W&B Team. The service account is created as a member of this team. For team-scoped service accounts, this is the only team they join. Organization-scoped service accounts are also added automatically to teams created later through SCIM. |
| `teams` | array | No | **Multi-tenant Cloud only.** Team names to add the account to. Include the same team as `defaultTeam` when you use this field. |

> #### Example

Reviewer: I'd suggest a more descriptive header and an intro sentence?
maybe 'Example requests for provisioning service accounts'

Author: This is a manually-maintained API reference effectively, and each section has an Example subsection: https://docs.wandb.ai/platform/hosting/iam/scim I think I will leave them all intact for now for consistency.

Reviewer: ah yes, consistency is a good reason to

<Tabs>
<Tab title="Provision Team Service Account Request">
```bash
POST /scim/Users
```

```json
{
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:teams:2.0:User"
    ],
    "userName": "sa-deploy-bot",
    "accountType": "SERVICE",
    "urn:ietf:params:scim:schemas:extension:teams:2.0:User": {
        "defaultTeam": "ml-platform"
    }
}
```
</Tab>
<Tab title="Provision Org Service Account Request">
```bash
POST /scim/Users
```

```json
{
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:teams:2.0:User"
    ],
    "userName": "sa-ci-runner",
    "accountType": "ORG_SERVICE",
    "urn:ietf:params:scim:schemas:extension:teams:2.0:User": {
        "defaultTeam": "ml-platform"
    }
}
```
</Tab>
</Tabs>

> #### Response

Reviewer: This feels odd as a section header to me?  seems like changing it to an intro sentence about 'the following response object is returned for both the team and org service account requests...

<Tabs>
<Tab title="Provision Service Account Response">
```bash
(Status 201)
```

```json
{
    "accountType": "SERVICE",
    "active": true,
    "displayName": "sa-deploy-bot",
    "id": "xyz",
    "meta": {
        "resourceType": "User",
        "created": "2023-10-01T00:00:00Z",
        "location": "Users/xyz"
    },
    "organizationRole": "member",
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:wandb:2.0:User"
    ],
    "teamRoles": [
        {
            "teamName": "ml-platform",
            "roleName": "member"
        }
    ],
    "urn:ietf:params:scim:schemas:extension:wandb:2.0:User": {
        "organizationRole": "member"
    },
    "userName": "sa-deploy-bot"
}
```
</Tab>
</Tabs>

For an organization-scoped service account, `accountType` is `ORG_SERVICE`.

> In **Self-Managed** deployments, `organizationRole` is `service` or `org_service` instead of `member`, matching the account type.

Reviewer: it feels like these two are input params to teh request? if so, feels like they should be up above the example requests (or at least directly after those examples), not here after the response.

Author: This is the HTTP code of the response. It belongs in the Response tab.

> A duplicate `userName` returns `409 Conflict`. A missing or unknown `defaultTeam` returns `400 Bad Request`.

Reviewer: errors returned from response? maybe an intro sentence and turning these two into bullets would help them scan better?  like 
"the response might return the following errors: 
 - a duplicate..
- A missing...

### Deprovision service account

Permanently deletes a service account and its organization membership. This is a hard delete; the account cannot be reactivated through SCIM.

<Note>
Available in **Dedicated Cloud** and **Self-Managed** v0.81.0+ and in **Multi-tenant Cloud**. Use the service account's SCIM user `id` from the provision response or from [Get user](#get-user). Deprovisioning does not delete API keys that were already issued; revoke keys separately in the W&B App if needed.
</Note>

#### Endpoint
- **URL**: `<host-url>/scim/Users/{id}`
- **Method**: DELETE

#### Parameters
| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|

> | `id` | string | Yes | The unique ID of the service account |

Reviewer: feels like this one maybe should ahve an ending period too since other tables on this same page had long sentences that ended in punctuation (consistency)

            "roleName": "member"
        }
    ],
    "urn:ietf:params:scim:schemas:extension:wandb:2.0:User": {
        "organizationRole": "member"
    },
    "userName": "sa-deploy-bot"
}
```
</Tab>
<Tab title="Provision Org Service Account Response">
```bash
(Status 201)
```

```json
{
    "accountType": "ORG_SERVICE",
    "active": true,
    "displayName": "sa-ci-runner",
    "id": "xyz",
    "meta": {
        "resourceType": "User",
        "created": "2023-10-01T00:00:00Z",
        "location": "Users/xyz"
    },
    "organizationRole": "member",
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User",
        "urn:ietf:params:scim:schemas:extension:wandb:2.0:User"
    ],
    "teamRoles": [
        {
            "teamName": "ml-platform",
            "roleName": "member"
        }
    ],
    "urn:ietf:params:scim:schemas:extension:wandb:2.0:User": {
        "organizationRole": "member"
    },
    "userName": "sa-ci-runner"
}
```
</Tab>
</Tabs>

For an organization-scoped service account, `accountType` is `ORG_SERVICE`.

In **Self-Managed** deployments, `organizationRole` is `service` or `org_service` instead of `member`, matching the account type.

> A duplicate `userName` returns `409 Conflict`. A missing or unknown `defaultTeam` returns `400 Bad Request`.

Author: ```suggestion
If the response returns one of these errors, check the request for these common problems:

- `409 Conflict`: The request includes duplicate `userName` keys for the same service account.
- `400 Bad Request`: The request is missing  `defaultTeam` or sets it to an invalid value.

```

### Deprovision service account

Permanently deletes a service account and its organization membership. This is a hard delete; the account cannot be reactivated through SCIM.

<Note>
Available in **Dedicated Cloud** and **Self-Managed** v0.81.0+ and in **Multi-tenant Cloud**. Use the service account's SCIM user `id` from the provision response or from [Get user](#get-user). Deprovisioning does not delete API keys that were already issued; revoke keys separately in the W&B App if needed.
</Note>

#### Endpoint
- **URL**: `<host-url>/scim/Users/{id}`
- **Method**: DELETE

#### Parameters
| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|

> | `id` | string | Yes | The unique ID of the service account |

Author: ```suggestion
| `id` | string | Yes | The unique ID of the service account. |
```
